UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentCallBackServiceTest.cs

```diff
@@ -0,0 +1,145 @@
+                    RequestedProductCount = 12,
+                    RequestedProductsAlreadyUpToDateCount = 5,
+                    RequestedProductsNotReturned = new List<RequestedProductsNotReturned>
+                    {
+                        new RequestedProductsNotReturned()
+                        {
+                            ProductName = "test",
+                            Reason = "notfound"
+                        }
+                    },
+                    ReturnedProductCount = 4
+                },
+                Products = new List<Products> {
+                            new Products {
+                                ProductName = "DE5NOBRK",
+                                EditionNumber = 0,
+                                UpdateNumbers = new List<int?> {0,1},
+                                FileSize = 400
+                            }
+                        }
+            };
+        }
+        #endregion
+
+        #region GetScsResponseQueueMessage
+        private SalesCatalogueServiceResponseQueueMessage GetScsResponseQueueMessage()
+        {
+            return new SalesCatalogueServiceResponseQueueMessage
+            {
+                BatchId = "7b4cdf10-adfa-4ed6-b2fe-d1543d8b7272",
+                FileSize = 4000,
+                ScsResponseUri = "https://test/ess-test/7b4cdf10-adfa-4ed6-b2fe-d1543d8b7272.json",
+                CallbackUri = "https://test-callbackuri.com",
+                CorrelationId = "727c5230-2c25-4244-9580-13d90004584a"
+            };
+        }
+        #endregion
+
+        [Test]
+        public async Task WhenValidCallBackUriInRequest_ThenSendCallBackResponseReturnsTrue()
+        {
+            A.CallTo(() => fakeCallBackClient.CallBackApi(A<HttpMethod>.Ignored, A<string>.Ignored, A<string>.Ignored))
+               .Invokes((HttpMethod method, string postBody, string uri) =>
+               {
+                   uriParam = uri;
+                   httpMethodParam = method;
+                   postBodyParam = postBody;
+               });
+
+            var response = await fulfilmentCallBackService.SendCallBackResponse(salesCatalogueProductResponse, scsResponseQueueMessage);
+
+            Assert.IsTrue(response);
+        }
+
+        [Test]
+        public async Task WhenEmptyCallBackUriInRequest_ThenSendCallBackResponseReturnsFalse()
+        {
+            scsResponseQueueMessage.CallbackUri = "";
+
+            A.CallTo(() => fakeCallBackClient.CallBackApi(A<HttpMethod>.Ignored, A<string>.Ignored, A<string>.Ignored))
+               .Invokes((HttpMethod method, string postBody, string uri) =>
+               {
+                   uriParam = uri;
+                   httpMethodParam = method;
+                   postBodyParam = postBody;
+               });
+
+            var response = await fulfilmentCallBackService.SendCallBackResponse(salesCatalogueProductResponse, scsResponseQueueMessage);
+
+            Assert.IsFalse(response);
+        }
+    }
```

Comment:
The purpose of the SendCallBackResponse method on the FulfilmentCallBackService is to prepare the payload and then call the CallBackApi on the callBackClient.

Would it possible to add a test to ensure this action takes place?

Reply:
As per the ESS API spec the service will ignore all response from the callback and hence we are not validating payload response.

Reply:
Hi  - if I commented out lines 40 to 68 on the FulfilmentCallBackService.cs the test would currently pass although the method wouldn't do anything.

In an ideal world we would aim to have sufficient tests to ensure that any failing code is detected by unit tests.

To do this a test could be added to check that callBackClient.CallBackApi is called.

Potentially with some refactoring tests could be added to ensure the payload is correctly created.

Reply:
as per the suggestion, i have added validation for request callback payload and added unit test case.